pkg/generate/cspc.go

```diff
@@ -0,0 +1,243 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generate
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	cstorv1 "github.com/openebs/api/v2/pkg/apis/cstor/v1"
+	"github.com/openebs/api/v2/pkg/apis/openebs.io/v1alpha1"
+	"github.com/openebs/openebsctl/pkg/client"
+	"github.com/openebs/openebsctl/pkg/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// isPoolValid checks if a CStor pool is valid
+func isPoolTypeValid(raid string) bool {
+	if raid == "stripe" || raid == "mirror" || raid == "raidz" || raid == "raidz2" {
+		return true
+	}
+	return false
+}
+
+// CSPC calls the generate routine for different cas-types
+func CSPC(nodes []string, devs int, raid string) error {
+	c := client.NewK8sClient()
+	if !isPoolTypeValid(strings.ToLower(raid)) {
+		// TODO: Use the well defined pool constant types from openebs/api when added there
+		return fmt.Errorf("invalid pool type %s", raid)
+	}
+	_, str, err := cspc(c, nodes, devs, strings.ToLower(raid))
+	if err != nil {
+		return err
+	}
+	fmt.Println(str)
+	return nil
+}
+
+// cspc takes eligible nodes, number of devices and poolType to create a pool cluster template
+func cspc(c *client.K8sClient, nodes []string, devs int, poolType string) (*cstorv1.CStorPoolCluster, string, error) {
+	// 0. Figure out the OPENEBS_NAMESPACE for CStor
+	cstorNS, err := c.GetOpenEBSNamespace(util.CstorCasType)
+	// assume CSTOR's OPENEBS_NAMESPACE has all the relevant blockdevices
+	c.Ns = cstorNS
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to determine the cStor namespace error: %v", err)
+	}
+	// 1. Validate nodes & poolType, fetch disks
+	nodeList, err := c.GetNodes(nodes, "", "")
+	if err != nil {
+		return nil, "", fmt.Errorf("(server error) unable to fetch node information %s", err)
+	}
+	if len(nodeList.Items) != len(nodes) {
+		return nil, "", fmt.Errorf("not all worker nodes are available for provisioning a cspc")
+	}
+	// 1.1 Translate nodeNames to node's hostNames to fetch disks
+	// while they might seem equivalent, they aren't equal, this quirk is
+	// visible clearly for EKS clusters
+	var hostnames []string
+	for _, node := range nodeList.Items {
+		// I hope it is unlikely for a K8s node to have an empty hostname
+		hostnames = append(hostnames, node.Labels["kubernetes.io/hostname"])
+	}
+	// 2. Fetch BD's from the eligible/valid nodes by hostname labels
+	bds, err := c.GetBDs(nil, "kubernetes.io/hostname in ("+strings.Join(hostnames, ",")+")")
+	if err != nil || len(bds.Items) == 0 {
+		return nil, "", fmt.Errorf("no blockdevices found in nodes with %v hostnames", hostnames)
```

Comment:
Good to log error also

Reply:
Returned errors are printed, the chain is a bit complex to the point it's not easy to figure these things out(which is not a good practice IMO) 😰 .

cspc.go:83 -> cspc.go:49 -> cmd/generate/generate.go:54 -> `util.CheckError is called with util.Fatal` and then this error is printed to stderr from pkg/util/util.go:74